### ISSUE TITLE
Saving session takes too much memory

### ISSUE THREAD

**Issue by CyberMew** [ProxymanApp/Proxyman#1111]
I was trying to save a session and it went as high as 80GB and I had to terminate it, losing all my data.

Is it possible to:

1. add cancellation in the progress window.
2. estimate the size required and check the disk space, prompt for warning first. then write to file in chunks periodically (newest records first) instead of saving all to memory. this way even if it crashes there is still something usable.

**Comment by NghiaTranUIT** [comment]
Do you remember what the latest ID number is? Or how long the session you've used? I would like to estimate how It can be generated as 80GB.

**Comment by CyberMew** [comment]
Sorry, unfortunately I don't remember at all, but for what's it worth my laptop has been up for 9 days and it tracks the network for my laptop and my phone. I was also surprised by the size the app was eating. Maybe there is a loop causing a leak somewhere? Regardless, I think it would be good to have point 1 and 2.

**Comment by CyberMew** [comment]
Since 21st Feb:
<img width="71" alt="CleanShot 2022-03-05 at 10 24 48@2x" src="https://user-images.githubusercontent.com/4715911/156864249-22b62a9d-68d0-48b2-9620-576c2bf8408f.png">
I would be happy to test it again, but I am pretty sure (like all my bug reports that have been valid so far) that this hang will happen again and the save will fail, and I need my requests to be around. Are you not able to replicate it on your side?

**Comment by NghiaTranUIT** [comment]
yes, I'm able to reproduce on my side. The problem is I could not find a better solution to reduce the RAM usage.

To clarify, the session is huge (150k items, and might be fully decrypted). Thus, it requires more RAM to serialize to JSON and zip it.

**Comment by NghiaTranUIT** [comment]
I highly suggest that you should only save a domain that you're working on. By right-clicking on the domain node on the left panel -> Export -> Proxyman Log, rather than trying to save entire the Session.

**Comment by CyberMew** [comment]
Glad you are able to replicate it. Having a progress window (UI not blocked) that shows the progress (while terminate-able) would be good. Processing it in limited chunks would be more manageable than dumping the whole thing to memory.

Thanks for the workaround suggestion, I will attempt to only export my pinned domain. May I know what is the difference between outputting as Proxyman Log vs Proxyman? And can it be restored in a future session of Proxyman if I quit the app?

**Comment by NghiaTranUIT** [comment]
Basically, ProxymanLog and ProxymanSession are the same. When you export it, you can import it later. 

ProxymanSession contains all domain/app nodes, so it generally is huge than ProxymanLog, which is for a particular node.

**Comment by CyberMew** [comment]
Just exported a domain and it basically went up to 40+GB of RAM using ProxymanLog. It was previously only 1GB+ iirc. Not sure how it can exceed 40GB+ but is it possible to make it such that it doesn't push the whole thing into memory eating into swap memory and instead do the processing piecewise? It makes the whole laptop lag when I export. I'd rather Proxyman process the export a bit slower and make the laptop usable, process cancel-able, with progress bar.

![CleanShot 2022-03-16 at 16 12 31@2x](https://user-images.githubusercontent.com/4715911/158545814-453453a6-327b-42a1-a7e9-a636fa9dcac5.png)

I also had to force quit the app because there is no way for me to cancel the operation and basically either lose the whole session or have all my other apps force close. As of now export is not very usable at all. Rather frustrating.

**Comment by matt-wood23** [comment]
+1 also seeing the same issue and unable to export my sessions, even when just exporting selected / filtered sessions. 

@NghiaTranUIT has there been any progress on resolving this and #1330 ? Both issues are many months old now.

**Comment by NghiaTranUIT** [comment]
Thanks @matt-wood23. We've recently received similar reports on the Export Session feature, which causes the app unresponsive.

I'm working on the v2 now and making sure it would consume less memory.

**Comment by matt-wood23** [comment]
Thanks @NghiaTranUIT - appreciate all the great work.
